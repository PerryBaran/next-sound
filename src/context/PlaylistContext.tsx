'use client'

import { createContext, useContext, useState } from 'react'

interface Song {
  image?: string
  artistName: string
  albumName: string
  songName: string
  audio: string
}

interface Playlist extends Song {
  key: string
}

const PlaylistContext = createContext({
  playlist: [] as Playlist[],
  playlistIndex: 0,
  skipSong: (i: number | boolean) => {},
  handleAddToPlaylist: (
    songs: Song[],
    addNext: boolean,
    playNow: boolean
  ) => {},
  removeFromPlaylist: (i: number) => {},
  playing: false,
  handlePlaying: (isPlaying: boolean | undefined) => {}
})

const usePlaylistContext = () => useContext(PlaylistContext)

function PlaylistProvider({ children }: { children: React.ReactNode }) {
  const [playlist, setPlaylist] = useState<Playlist[]>([])
  const [playlistIndex, setPlaylistIndex] = useState(0)
  const [playing, setPlaying] = useState(false)

  const addToPlaylist = (songs: Song[], addNext = false) => {
    const data = songs.map((song) => {
      return { ...song, key: crypto.randomUUID() }
    })

    setPlaylist((prev) => {
      if (addNext) {
        const clone = [...prev]
        clone.splice(playlistIndex + 1, 0, ...data)
        return clone
      }
      return [...prev, ...data]
    })
  }

  const handlePlayNow = () => {
    if (playlist.length !== 0) {
      setPlaylistIndex((prev) => prev + 1)
    }
    setPlaying(true)
  }

  const handleAddToPlaylist = (
    songs: Song[],
    addNext: boolean,
    playNow: boolean
  ) => {
    addToPlaylist(songs, addNext)
    if (playNow) {
      handlePlayNow()
    }
  }

  const removeFromPlaylist = (i: number) => {
    setPlaylist((prev) => {
      const clone = [...prev]
      clone.splice(i, 1)
      return clone
    })

    setPlaylistIndex((prev) => {
      if (i > prev || prev === 0) {
        return prev
      }
      return prev - 1
    })
  }

  const skipSong = (value: number | boolean) => {
    const lastPlaylistIndex = playlist.length - 1
    if (typeof value === 'number') {
      if (value > lastPlaylistIndex) {
        setPlaylistIndex(lastPlaylistIndex)
      } else if (value < 0) {
        setPlaylistIndex(0)
      } else {
        setPlaylistIndex(value)
      }
    } else if (value) {
      if (playlistIndex === lastPlaylistIndex) {
        setPlaylistIndex(0)
      } else {
        setPlaylistIndex((prev) => prev + 1)
      }
    } else {
      if (playlistIndex > 0) {
        setPlaylistIndex((prev) => prev - 1)
      } else {
        setPlaylistIndex(lastPlaylistIndex)
      }
    }
  }

  const handlePlaying = (isPlaying: boolean | undefined) => {
    if (isPlaying === undefined) {
      setPlaying((prev) => !prev)
    } else {
      setPlaying(isPlaying)
    }
  }

  return (
    <PlaylistContext.Provider
      value={{
        playlist,
        playlistIndex,
        handleAddToPlaylist,
        removeFromPlaylist,
        skipSong,
        playing,
        handlePlaying
      }}
    >
      {children}
    </PlaylistContext.Provider>
  )
}

export { usePlaylistContext, PlaylistProvider }
