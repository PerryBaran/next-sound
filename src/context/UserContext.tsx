'use client'

import { createContext, useContext, useState, useEffect } from 'react'
import { getUserById } from '@/requests/users'
import Cookie from 'js-cookie'
import jwtDecode from 'jwt-decode'

interface User {
  name?: string
  id?: string
}

const emptyUser = {
  name: '',
  id: ''
}

const UserContext = createContext({
  user: emptyUser,
  handleLogin: (data: User | undefined) => {}
})

const useUserContext = () => useContext(UserContext)

function UserProvider({ children }: { children: React.ReactNode }) {
  const [user, setUser] = useState(emptyUser)

  const handleLogin = (data: User | undefined) => {
    if (data === undefined) {
      Cookie.remove('userToken')
      setUser(emptyUser)
    } else {
      setUser((prev) => {
        return { ...prev, ...data }
      })
    }
  }

  useEffect(() => {
    ;(async () => {
      const token = Cookie.get('userToken')
      if (!token) return
      const { id }: { id: string } = jwtDecode(token)
      const response = await getUserById(id)
      handleLogin({
        name: response.name,
        id
      })
    })()
  }, [])

  return (
    <UserContext.Provider value={{ user, handleLogin }}>
      {children}
    </UserContext.Provider>
  )
}

export { useUserContext, UserProvider }
