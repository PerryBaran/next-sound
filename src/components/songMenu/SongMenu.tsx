"use client"

import { useState } from "react"
import css from "./songMenu.module.css"
import { usePlaylistContext } from "../../context/PlaylistContext"
import { play, addList, dots } from "../../media/icons/index"
import Image from "next/image"

interface Props {
  songs: {
    songName: string
    audio: string
    image?: string
    artistName: string
    albumName: string
  }[]
}

export default function SongMenu({ songs }: Props) {
  const { handleAddToPlaylist } = usePlaylistContext()
  const [visible, setVisible] = useState(false)

  const handleBlur = () => {
    setVisible(false)
  }

  return (
    <div className={css.container} onBlur={handleBlur}>
      {visible ? (
        <div className={css.options}>
          <button onMouseDown={() => handleAddToPlaylist(songs, true, true)}>
            <span className={css["song-menu-image-container"]}>
              <Image src={play} alt="play now" width={9} />
            </span>
            <span>Play Now</span>
          </button>
          <button
            onMouseDown={() => handleAddToPlaylist(songs, true, false)}
            className={css["play-next"]}
          >
            <span className={css["song-menu-image-container"]}>
              <Image src={addList} alt="play now" width={15} />
            </span>
            <span>Play Next</span>
          </button>
          <button onMouseDown={() => handleAddToPlaylist(songs, false, false)}>
            <span className={css["song-menu-image-container"]}>
              <Image src={addList} alt="play now" width={15} />
            </span>
            <span>Add To Queue</span>
          </button>
        </div>
      ) : null}
      <button
        onClick={() => setVisible((prev) => !prev)}
        className={css.button}
      >
        <Image src={dots} alt="..." height={18} width={18} />
      </button>
    </div>
  )
}
