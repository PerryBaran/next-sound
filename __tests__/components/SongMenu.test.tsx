import { fireEvent, render, screen } from "@testing-library/react"
import { createContext, useContext } from "react"
import SongMenu from "../../src/components/songMenu/SongMenu"
import * as PlaylistContext from "../../src/context/PlaylistContext"

describe("SongMenu", () => {
  const props = {
    songs: [
      {
        songName: "song-1",
        audio: "audio-1",
        image: "/image-1",
        artistName: "artist-1",
        albumName: "album-1"
      },
      {
        songName: "song-2",
        audio: "audio-2",
        image: "/image-2",
        artistName: "artist-2",
        albumName: "album-2"
      }
    ]
  }
  let mockHandleAddToPlaylist = jest.fn()

  beforeEach(() => {
    mockHandleAddToPlaylist = jest.fn()
    const mockContext = createContext({
      playlist: [],
      playlistIndex: 0,
      skipSong: jest.fn(),
      handleAddToPlaylist: mockHandleAddToPlaylist,
      removeFromPlaylist: jest.fn(),
      playing: false,
      handlePlaying: jest.fn(),
      handleShuffle: jest.fn()
    })

    jest
      .spyOn(PlaylistContext, "usePlaylistContext")
      .mockImplementation(() => useContext(mockContext))
  })

  test("renders correctly", () => {
    render(<SongMenu {...props} />)

    expect(screen.getByRole("button")).toBeInTheDocument()
    expect(screen.getByAltText("...")).toBeInTheDocument()
  })

  test("on button click shows more options, on second click hides options", () => {
    render(<SongMenu {...props} />)

    fireEvent.click(screen.getByAltText("..."))

    expect(screen.getAllByRole("button")).toHaveLength(4)
    expect(screen.getByText(/play now/i)).toBeInTheDocument()
    expect(screen.getByText(/play next/i)).toBeInTheDocument()
    expect(screen.getByText(/add to queue/i)).toBeInTheDocument()

    fireEvent.click(screen.getByAltText("..."))

    expect(screen.getAllByRole("button")).toHaveLength(1)
  })

  test("on container blur hides options", () => {
    render(<SongMenu {...props} />)

    fireEvent.click(screen.getByAltText("..."))

    expect(screen.getAllByRole("button")).toHaveLength(4)

    const container = screen.getByAltText("...").parentElement
    if (!container) throw new Error("missing container")
    fireEvent.blur(container)

    expect(screen.getAllByRole("button")).toHaveLength(1)
  })

  test("play now mouseDown", () => {
    render(<SongMenu {...props} />)

    fireEvent.click(screen.getByAltText("..."))
    fireEvent.mouseDown(screen.getByText(/play now/i))

    expect(mockHandleAddToPlaylist).toBeCalledTimes(1)
    expect(mockHandleAddToPlaylist).toBeCalledWith(props.songs, true, true)
  })

  test("play next mouseDown", () => {
    render(<SongMenu {...props} />)

    fireEvent.click(screen.getByAltText("..."))
    fireEvent.mouseDown(screen.getByText(/play next/i))

    expect(mockHandleAddToPlaylist).toBeCalledTimes(1)
    expect(mockHandleAddToPlaylist).toBeCalledWith(props.songs, true, false)
  })

  test("add to queue mouseDown", () => {
    render(<SongMenu {...props} />)

    fireEvent.click(screen.getByAltText("..."))
    fireEvent.mouseDown(screen.getByText(/add to queue/i))

    expect(mockHandleAddToPlaylist).toBeCalledTimes(1)
    expect(mockHandleAddToPlaylist).toBeCalledWith(props.songs, false, false)
  })

  test("snapshot", () => {
    const { asFragment } = render(<SongMenu {...props} />)

    expect(asFragment()).toMatchSnapshot()
  })
})
