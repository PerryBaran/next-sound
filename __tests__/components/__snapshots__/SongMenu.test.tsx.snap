// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SongMenu snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <button
      class="button"
    >
      <img
        alt="..."
        data-nimg="1"
        decoding="async"
        height="18"
        loading="lazy"
        src="/_next/image?url=%2Fimg.jpg&w=48&q=75"
        srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=48&q=75 2x"
        style="color: transparent;"
        width="18"
      />
    </button>
  </div>
</DocumentFragment>
`;
